src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/KestrelThread.cs

```diff
@@ -93,28 +93,33 @@ public void Stop(TimeSpan timeout)
                 }
             }
 
+            var stepTimeout = (int)(timeout.TotalMilliseconds / 3);
```

Comment:
Actually, this probably isn't what we want, since the whole timeout could be exhausted by the first call to `Join()`.  I would probably just pass the `timeout` parameter directly to each call to `Join()`, and reduce the timeout passed by the caller if you want to lower the total maximum time.